packages/terminal/src/browser/terminal-quick-open-service.ts

```diff
@@ -0,0 +1,155 @@
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { inject, injectable } from 'inversify';
+import {
+    QuickOpenModel, QuickOpenItem, QuickOpenHandler,
+    QuickOpenOptions, QuickOpenItemOptions, QuickOpenMode,
+    PrefixQuickOpenService,
+    QuickOpenContribution, QuickOpenHandlerRegistry
+} from '@theia/core/lib/browser';
+import { CommandContribution } from '@theia/core/lib/common';
+import { CommandRegistry } from '@theia/core/lib/common';
+import { TerminalWidget } from './base/terminal-widget';
+import { TerminalService } from './base/terminal-service';
+import { TerminalCommands } from './terminal-frontend-contribution';
+
+@injectable()
+export class TerminalQuickOpenService implements QuickOpenModel, QuickOpenHandler {
+
+    @inject(PrefixQuickOpenService)
+    protected readonly prefixQuickOpenService: PrefixQuickOpenService;
+
+    @inject(CommandRegistry)
+    protected readonly commandRegistry: CommandRegistry;
+
+    @inject(TerminalService)
+    protected readonly terminalService: TerminalService;
+
+    readonly prefix: string = 'term ';
+
+    get description(): string {
+        return 'Show All Opened Terminals';
+    }
+
+    getModel(): QuickOpenModel {
+        return this;
+    }
+
+    getOptions(): QuickOpenOptions {
+        return {
+            fuzzyMatchLabel: {
+                enableSeparateSubstringMatching: true
+            },
+            fuzzyMatchDescription: {
+                enableSeparateSubstringMatching: true
+            }
+        };
+    }
+
+    open(): void {
+        this.prefixQuickOpenService.open(this.prefix);
+    }
+
+    async onType(lookFor: string, acceptor: (items: QuickOpenItem[]) => void): Promise<void> {
+        const terminalItems: QuickOpenItem[] = [];
+
+        // Get the list of currently opened terminal widgets, sorted by the widget id.
+        const widgets: TerminalWidget[] = this.terminalService.all
+            .sort((a: TerminalWidget, b: TerminalWidget) => (a.id.localeCompare(b.id)));
+
+        for (const widget of widgets) {
+            const item = await this.toItem(widget);
+            terminalItems.push(item);
+        }
+        // Append a quick open item to create a new terminal.
+        const createNewTerminalItem = new QuickOpenItem({
+            label: 'Open New Terminal',
```

Comment:
Do you think we should add an icon class for the create?

```ts
iconClass: 'fa fa-plus'
```

Reply:
Sure.